pints/_mcmc/_nuts.py

```diff
@@ -0,0 +1,557 @@
+        self.theta_minus = theta
+        self.theta_plus = theta
+        self.r_minus = r
+        self.r_plus = r
+        self.L_minus = L
+        self.L_plus = L
+        self.grad_L_minus = grad_L
+        self.grad_L_plus = grad_L
+        self.n = n
+        self.s = s
+        self.theta = theta
+        self.L = L
+        self.grad_L = grad_L
+        self.alpha = alpha
+        self.n_alpha = n_alpha
+
+    def update(self, other_state, direction, root):
+        """
+        if root == True, this combines a depth j subtree (`self`) with a depth j+1
+        (`other_state`) subtree, which corresponds to the higher level loop in
+        the nuts algorithm
+
+        if root == False, this combins two subtrees with depth j, which occurs
+        when the nuts algorithm is implicitly building up the tree with the build_tree
+        subroutine
+
+        direction is the current direction of integration, either forwards
+        (direction == 1), or backwards (direction = -1)
+        """
+        if direction == -1:
+            self.theta_minus = other_state.theta_minus
+            self.r_minus = other_state.r_minus
+            self.L_minus = other_state.L_minus
+            self.grad_L_minus = other_state.grad_L_minus
+        else:
+            self.theta_plus = other_state.theta_plus
+            self.r_plus = other_state.r_plus
+            self.L_plus = other_state.L_plus
+            self.grad_L_plus = other_state.grad_L_plus
+
+        theta_dash = other_state.theta
+        L_dash = other_state.L
+        grad_L_dash = other_state.grad_L
+        n_dash = other_state.n
+        s_dash = other_state.s
+        alpha_dash = other_state.alpha
+        n_alpha_dash = other_state.n_alpha
+
+        # if there is any accepted points in the other subtree then test for acceptance
+        # of that subtrees theta
+        if n_dash > 0:
+            if root:
+                p = int(s_dash == 1)*min(1, n_dash / self.n)
+            else:
+                p = n_dash / (self.n + n_dash)
+
+            if p > 0.0 and np.random.uniform() < p:
+                self.theta = theta_dash
+                self.L = L_dash
+                self.grad_L = grad_L_dash
+
+        # Nots: alpha and n_alpha are only accumulated within build_tree
+        if root:
+            self.alpha = alpha_dash
+            self.n_alpha = n_alpha_dash
+        else:
+            self.alpha += alpha_dash
+            self.n_alpha += n_alpha_dash
+
+        # accumulated number of accepted points
+        self.n += n_dash
+
+        # test if the path has done a U-Turn
+        self.s *= s_dash
+        self.s *= int((self.theta_plus - self.theta_minus).dot(self.r_minus) >= 0)
+        self.s *= int((self.theta_plus - self.theta_minus).dot(self.r_plus) >= 0)
+
+
+@asyncio.coroutine
+def leapfrog(theta, L, grad_L, r, epsilon, step_size):
+    """ performs a leapfrog step with step size `epsilon*step_size """
```

Comment:
`Performs` and ` ``epsilon*step_size``. `

Reply:
done